File: java-workspace2/src/복습/정수삼각형.java
package 복습;

import java.util.Scanner;

public class 정수삼각형 {
    static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        int maxFloor = scanner.nextInt();
        int answer = 0;
        int[][] costs = new int[maxFloor + 1][maxFloor + 1];
        int[][] dp = new int[maxFloor + 1][maxFloor + 1];
        for (int floor = 1; floor <= maxFloor; floor++) {
            for (int idx = 1; idx <= floor; idx++) {
                costs[floor][idx] = scanner.nextInt();
            }
        }
        for (int floor = 1; floor <= maxFloor; floor++) {
            for (int idx = 1; idx <= maxFloor; idx++) {
                dp[floor][idx] = Math.max(dp[floor - 1][idx] + costs[floor][idx],
                        dp[floor - 1][idx - 1] + costs[floor][idx]);
            }
        }
        for (int idx = 1; idx <= maxFloor; idx++) {
            answer = Math.max(answer, dp[maxFloor][idx]);
        }
        System.out.println(answer);
    }
}

File: python-workspace/백준/다이나믹프로그래밍/정수삼각형.py
import sys

max_floor = int(sys.stdin.readline())
cost = [[0 for _ in range(max_floor + 1)] for _ in range(max_floor + 1)]
for floor in range(1, max_floor + 1):
    for idx in range(1, max_floor + 1):
        cost[floor][idx] = int(input())
print(cost)
for _ in range(max_floor):
    cost.append(list(map(int, sys.stdin.readline().split(" "))))
dp = [[0 for _ in range(max_floor + 1)] for _ in range(max_floor + 1)]

for floor in range(1, max_floor + 1):
    for idx in range(1, max_floor + 1):
        print("floor " + str(floor) + " idx " + str(idx))
        dp[floor][idx] = max(dp[floor - 1][idx - 1] + cost[floor][idx], dp[floor - 1][idx])

answer = None
for candidate in dp[max_floor]:
    answer = max(answer, candidate)
print(answer)
